pyiceberg/table/__init__.py

```diff
@@ -484,10 +484,6 @@ def append(self, df: pa.Table, snapshot_properties: Dict[str, str] = EMPTY_DICT)
         _check_schema_compatible(
             self._table.schema(), other_schema=df.schema, downcast_ns_timestamp_to_us=downcast_ns_timestamp_to_us
         )
-        # cast if the two schemas are compatible but not equal
```

Comment:
@syun64 I want to get your take on this part. Due to the timestamp change, do you know if the `df` need to be casted? 
There are a couple of different parts involved in the write path. In particular, we need to look at the table schema, the df schema, and the df itself. As well as dealing with bin-packing and other transformations.

Reply:
Happy to extract this convo into an issue, to also continue the convo from https://github.com/apache/iceberg-python/pull/786#discussion_r1646417180

Reply:
> @syun64 I want to get your take on this part. Due to the timestamp change, do you know if the `df` need to be casted? There are a couple of different parts involved in the write path. In particular, we need to look at the table schema, the df schema, and the df itself. As well as dealing with bin-packing and other transformations.

I have a PR open to try to fix this behavior: https://github.com/apache/iceberg-python/pull/910 I think it's almost ready to merge 😄